### PR TITLE
Improve `LazyList` self-reference error

### PR DESCRIPTION
**Pull request by NthPortal**
Improve error message and docs for `LazyList` self-reference error.

Happy to bikeshed

### CODE REVIEW

File: test/junit/scala/collection/immutable/LazyListTest.scala [hunk 1]
```diff
@@ -11,7 +11,6 @@ import scala.collection.mutable.{Builder, ListBuffer}
 import scala.tools.testkit.{AssertUtil, ReflectUtil}
 import scala.util.Try
 
-@RunWith(classOf[JUnit4])
```
Comment (som-snytt) [inline test/junit/scala/collection/immutable/LazyListTest.scala:14]: The imports are unused now. We punish scouting with more scouting. That's the CI failure.

File: src/library/scala/collection/immutable/LazyList.scala [hunk 2]
```diff
@@ -253,7 +274,9 @@ final class LazyList[+A] private(private[this] var lazyState: () => LazyList.Sta
     // if it's already mid-evaluation, we're stuck in an infinite
     // self-referential loop (also it's empty)
     if (midEvaluation) {
-      throw new RuntimeException("self-referential LazyList or a derivation thereof has no more elements")
+      throw new RuntimeException(
+        "LazyList evaluation depends on its own result (self-reference); see docs for more info"
+        )
```
Comment (som-snytt) [inline src/library/scala/collection/immutable/LazyList.scala]: I recently inherited a more modern monitor, so this looks short enough for one line. But I did not count chars.

File: src/library/scala/collection/immutable/LazyList.scala [hunk 3]
```diff
@@ -218,6 +218,27 @@ import scala.runtime.Statics
...
+  *  trivial example of such might be
+  *
+  *  {{{
+  *  lazy val a: LazyList[Int] = 1 #:: 2 #:: a.filter(_ > 2)
+  *  }}}
+  *
+  *  When attempting to evaluate the third element of `a`, it will skip the first two
+  *  elements and read the third, but that element is already being evaluated. This is
+  *  often caused by a subtle logic error; in this case, using `>=` in the `filter`
+  *  would fix the error.
```
Comment (som-snytt) [inline src/library/scala/collection/immutable/LazyList.scala:240]: That is very subtle. I will have to fully context-switch to understand it. At the moment, I see from the previous exception message that it has something to do with "no more elements". To borrow a phrase, it's amazing that it works at all.

Comment (He-Pin) [reply src/library/scala/collection/immutable/LazyList.scala:240]: Better provides a general guide section?

Comment (NthPortal) [reply src/library/scala/collection/immutable/LazyList.scala:240]: do you think I should keep "no more elements" in the message?

File: src/library/scala/collection/immutable/LazyList.scala (hunk 2 above)
Comment (NthPortal) [reply src/library/scala/collection/immutable/LazyList.scala]: after my change, part of the last character and closing quote run off the end of the window. on the other hand, I use an unusually large font size. on the other other hand, we should try to accommodate as many people as possible, so erring shorter is maybe better anyway?

in its current state in the PR, the line has 97 characters. if I put the message on the same line as the constructor, it has 123

File: src/library/scala/collection/immutable/LazyList.scala (hunk 3 above)
Comment (NthPortal) [reply src/library/scala/collection/immutable/LazyList.scala:240]: tbh, I'm not happy with my docs either, but they seem better than just the previous message hopefully? and can be incrementally improved by others as well

**Review by lrytz — APPROVED** [review]
Thank you!

**Review by SethTisue — APPROVED** [review]

**Comment by NthPortal** [comment]
oops, sorry I didn't get back to this for the cleanup. life has been hectic

**Comment by SethTisue** [comment]
@NthPortal is "Marissa" how you want to show up in the release notes? if not, please submit a PR updating https://github.com/scala/scala/blob/2.13.x/.mailmap appropriately

**Comment by NthPortal** [comment]
> is "Marissa" how you want to show up in the release notes?

yes. although I don't actually see it in the mailmap, but I will take your word for it that I don't need to do anything

**Pull request merged**